File: src/main/java/io/sch/historyscan/domain/contexts/analysis/history/CodebaseHistoryAnalysis.java
package io.sch.historyscan.domain.contexts.analysis.history;

import io.sch.historyscan.domain.contexts.analysis.common.Analyze;
import io.sch.historyscan.domain.contexts.analysis.common.CodeBaseToAnalyze;
import io.sch.historyscan.domain.error.HistoryScanFunctionalException;

public class CodebaseHistoryAnalysis implements Analyze<CodeBaseHistory> {

    private final HistoryAnalyzer historyAnalyzer;

    public CodebaseHistoryAnalysis(HistoryAnalyzer historyAnalyzer) {
        this.historyAnalyzer = historyAnalyzer;
    }

    @Override
    public CodeBaseHistory from(CodeBaseToAnalyze codeBaseToAnalyze) throws HistoryScanFunctionalException {
        return historyAnalyzer.of(codeBaseToAnalyze.getName())
                .orElseThrow(() ->
                        new CodeBaseHistoryNotFoundException(
                                "CodeBase '%s' history not found".formatted(codeBaseToAnalyze.getName())));
    }
}

File: historyscan-back/src/main/kotlin/io/sch/historyscan/domain/contexts/analysis/history/CodebaseHistoryAnalysis.kt
package io.sch.historyscan.domain.contexts.analysis.history

import io.sch.historyscan.domain.contexts.analysis.common.*
import io.sch.historyscan.domain.error.HistoryScanFunctionalException

class CodebaseHistoryAnalysis(private val historyAnalyzer: HistoryAnalyzer) : Analyze<CodeBaseHistory> {

    @Throws(HistoryScanFunctionalException::class)
    override fun from(codeBaseToAnalyze: CodeBaseToAnalyze): CodeBaseHistory {
        return historyAnalyzer.of(codeBaseToAnalyze.name)
            ?: throw CodeBaseHistoryNotFound("Unable to get codebase history")
    }
}
